### ISSUE TITLE
"Updating" images in Word document ?

### ISSUE THREAD

**Issue by behrica** [davidgohel/officer#81]
My use case is, that I collaborate with colleges on a word document,
and my task is to generate "plots" (= images).

So what I want to do repeatedly is to "update" a word document  automatically,
replacing "all images" (or some at least)  with a new version of the image.

I have read the documentation about using "body_remove" 
"cursor_bookmarks" and others, but I am still wondering if this can be done reliably with existing officer functions.

I still feel, that something is missing for this use case.

I looked at the xml representation of a paragraph, which contains an image,
and I have somehow the impression that "update an image" is not really foreseen by the existing function in officer. (or at least "brittle")

I start to think that the best way for my use case is to simply "replace" the images files
in the "word/media" folder with new versions.

Maybe a function in officer would help, which allows to "extract" all "image links"
so, an xpath like "w:drawing/wp:docPr" which extracts from the docx the relation between a "w:drawing" and the "name of the image file in the docx zip (in "media").

And then I would be rather easy to just "copy" a new version of the images diretcly into the zip. This would the "update" the images. (without doing a "remove" and "add" image operation).

I think in officer there is some code, which helps with this, as you have functions to "add images", so to create the relationships.
I saw function "officer:::wml_link_images", which does the opposite.

So ideally I would like to see two function added in officer:

"docx_images(docx) ", which returns "some data" on all images.
This should "show" the relationship between each w:drawing and the "image file" in the  "media" folder.

"update_image_file(image_id, fileName)"  or similar.

Looking at the xml, maybe the following "data" on images could be of interest:
<wp:docPr id="21" name="Picture 21"/>
.....
<pic:nvPicPr>
                                    <pic:cNvPr id="0" name="locationsDensity.png"/>
                                    <pic:cNvPicPr/>
                                </pic:nvPicPr>

**Comment by behrica** [comment]
I saw that the function to list images is basically there:

 doc$doc_obj$relationship()$get_data()

What this function returns, would miss for my use case a key information,
namely "which image" is this. All theses ids it gives are very abstract.

I saw that the xml, contains this piece of data:
```xml
<pic:nvPicPr>
            <pic:cNvPr id="0" name="logo_RGB.png"/>
            <pic:cNvPicPr/>
</pic:nvPicPr>
```

I believe that the "name" in here, is the "filename" used, when an image get inserted "manualy" into a word document.

If the "getData()" method would return this piece as well, this could hel me to decide with file in "media/" to update...

**Comment by davidgohel** [comment]
I don't think using bookmarks is *brittle*. 

Can you show the code you are using that make you feel that a new function is necessary?

**Comment by behrica** [comment]
With brittle I mean,
that to use the bookmark I needed to:

1. trust that my colaborators don't remove the bookmarks... 
2. I would not be sure where to place the bookmark in the first place ("on" the image, before, after)
3. Then "body_remove" would "blindly" remove something which is before the bookmark (whatever it is) (hoping it to be still the image)
4. This might then "remove the bookmark itself"...

it is true, that I have not really tried it, just read the documentation of the available functions.

Given my use case (repeated replacement of images),
what exactly would you suggest to "bookmark"
(the image, the image caption text, the paragrpah before the image, or after ?)

And which function (add_img) vs slip_in should be used ?

I am indeed wondering, if I should "replace" the paragraph (including) the image or "replace" the image inside the same paragraph....

**Comment by davidgohel** [comment]
I suggest to have bookmarks just before images you want to replace. Then to replace the following paragraphs using `body_remove` and `body_add_img`.

If you found a better solution to replace images in Word documents, submit a PR. I don't have time to create a specific function for your need but I will try to find some next year (this is mainly developed in my spare time).

**Comment by vinhdizzo** [comment]
Hi,

I followed @davidgohel's advice on bookmarking the image location, and removing then adding image to "update" images in the doc. However, similar @behrica, I can't get it down correctly. Should the bookmark be on the image? On the line before? On the line after?

I tried this on a doc with 2 images, 2 bookmarks, where bookmark is on line before image. However, bookmarks seem to get deleted when I remove body. I also tried moving the cursor next, but can't get things down correctly. Any help would be appreciated.

I will note that I could do this without problems on the pptx front, as I could determine slide I want, then image, then delete, then add. Docx is a little harder as it is tricky to determine where the image is exactly.
